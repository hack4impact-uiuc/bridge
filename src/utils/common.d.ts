declare module "@hack4impact-uiuc/bridge/common" {
  import { Props, ReactType } from "react";
  import {
    ColorProps,
    SpaceProps,
    LayoutProps,
    TypographyProps as StyledTypographyProps,
    BordersProps as StyledBordersProps,
    BoxShadowProps,
  } from "styled-system";

  export interface BaseProps extends Props<any> {
    as?: ReactType;
    className?: string;
  }
  export interface CommonProps
    extends BaseProps,
      ColorProps,
      SpaceProps,
      LayoutProps {}

  export interface TypographyProps extends BaseProps, StyledTypographyProps {}

  export interface BorderProps extends StyledBordersProps, BoxShadowProps {}
}
