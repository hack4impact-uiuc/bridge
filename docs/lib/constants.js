export const COMPONENTS_URL = '/components';
export const DESIGN_URL = '/design';
