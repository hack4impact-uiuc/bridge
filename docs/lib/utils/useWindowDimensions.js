import { useState, useEffect } from 'react';

function getWindowDimensions() {
  const { innerWidth: width, innerHeight: height } = window;
  return {
    width,
    height,
  };
}

export default function useWindowDimensions() {
  const [windowDimensions, setWindowDimensions] = useState(getWindowDimensions());

  useEffect(() => {
    function handleResize() {
      const { innerWidth: width, innerHeight: height } = window;
      setWindowDimensions({ width, height });
    }

    window.addEventListener('resize', handleResize);
    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return windowDimensions;
}
