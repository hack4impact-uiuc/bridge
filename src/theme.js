/* eslint-disable prefer-destructuring */

function fontStack(fonts) {
  return fonts.map((font) => (font.includes(' ') ? `"${font}"` : font)).join(', ');
}

const DEFAULT_SHADOW = '0 1px 6px 2px rgba(0,0,0,0.15)';

const baseColors = {
  hackBlue: ['#B4D8FA', '#3E87CD', '#155DA1', '#0E4E8A', '#0D3F6E', '#031425'],
  indigoBlue: ['#BCD0F3', '#415F94', '#2D4979', '#253E68', '#13294E', '#061734'],
  impactGrey: ['#FAFBFC', '#EBEEF2', '#8B9199', '#666B72', '#46494F'],
};

const hackBluePalette = {
  lighter: baseColors.hackBlue[0],
  light: baseColors.hackBlue[1],
  primary: baseColors.hackBlue[2],
  dark: baseColors.hackBlue[3],
  darker: baseColors.hackBlue[4],
  text: baseColors.hackBlue[5],
};

const indigoPalette = {
  lighter: baseColors.indigoBlue[0],
  light: baseColors.indigoBlue[1],
  primary: baseColors.indigoBlue[2],
  dark: baseColors.indigoBlue[3],
  darker: baseColors.indigoBlue[4],
  text: baseColors.indigoBlue[5],
};

const impactGreyPalette = {
  lighter: baseColors.impactGrey[0],
  light: baseColors.impactGrey[1],
  midtone: baseColors.impactGrey[2],
  primary: baseColors.impactGrey[3],
  darker: baseColors.impactGrey[4],
};

const secondaryColors = {
  green: ['#BDF3C6', '#90E19E', '#4C9859', '#256830', '#154C1E', '#031D07'],
  red: ['#FFBCBF', '#D1595F', '#A62B31', '#8A1A1F', '#6B1014', '#35080A'],
  yellow: ['#FFECBD', '#FFD15F', '#FDBE21', '#DA9C03', '#A77700', '#201803'],
};

const greenPalette = {
  lighter: secondaryColors.green[0],
  light: secondaryColors.green[1],
  primary: secondaryColors.green[2],
  dark: secondaryColors.green[3],
  darker: secondaryColors.green[4],
  text: secondaryColors.green[5],
};

const redPalette = {
  lighter: secondaryColors.red[0],
  light: secondaryColors.red[1],
  primary: secondaryColors.red[2],
  dark: secondaryColors.red[3],
  darker: secondaryColors.red[4],
  text: secondaryColors.red[5],
};

const yellowPalette = {
  lighter: secondaryColors.yellow[0],
  light: secondaryColors.yellow[1],
  primary: secondaryColors.yellow[2],
  dark: secondaryColors.yellow[3],
  darker: secondaryColors.yellow[4],
  text: secondaryColors.yellow[5],
};

const typographyColors = {
  charcoal: '#272A2D',
  iron: '#4B4E52',
  ink: '#0A162A',
  slate: '#363949',
};

const colors = {
  black: '#000000',
  white: '#FFFFFF',
  ...baseColors,
  ...secondaryColors,
  ...typographyColors,
  hackBluePalette,
  indigoPalette,
  impactGreyPalette,
  greenPalette,
  redPalette,
  yellowPalette,
  text: {
    grayScale: {
      main: typographyColors.charcoal,
      support: typographyColors.iron,
    },
    coolTone: {
      main: typographyColors.ink,
      support: typographyColors.slate,
    },
  },
  // states or in other words color variants
  variants: {
    primary: hackBluePalette,
    success: greenPalette,
    error: redPalette,
    warning: yellowPalette,
    secondary: hackBluePalette,
    white: {
      primary: '#FFFFFF',
      dark: '#FFFFFF',
      lighter: indigoPalette.light,
    },
  },
};

const fontSizes = ['12px', '14px', '16px', '18px', '20px', '22px', '25px', '28px', '32px', '45px', '60px'];

const fonts = {
  chivo: fontStack(['Chivo']),
  hkGrotesk: fontStack(['HKGrotesk']),
};

const fontWeights = {
  light: 300,
  normal: 400,
  medium: 500,
  bold: 700,
};

// styling specific to the Button component
const button = {
  font: fonts.chivo,
  fontWeight: fontWeights.bold,
  fontSize: '14px',
  letterSpacing: '2px',
  // default values, based on the color palette
  default: (colorPalette) => ({
    bg: {
      default: colorPalette.primary,
      hover: colorPalette.dark,
      disabled: colorPalette.lighter,
    },
    border: {
      default: 'none',
      hover: 'none',
      disabled: 'none',
    },
    fontColor: {
      default: colors.white,
      hover: colors.white,
      disabled: colorPalette.light,
    },
    shadow: {
      default: 'none',
      hover: 'none',
      disabled: 'none',
    },
    outline: {
      bg: {
        default: colors.white,
        hover: colors.white,
        disabled: colors.white,
      },
      border: {
        default: colorPalette.primary,
        hover: colorPalette.dark,
        disabled: colorPalette.lighter,
      },
      fontColor: {
        default: colorPalette.primary,
        hover: colorPalette.dark,
        disabled: colorPalette.lighter,
      },
      shadow: {
        default: 'none',
        hover: 'none',
        disabled: 'none',
      },
    },
  }),

  // everything below overrides the default values
  primary: {
    fontColor: {
      default: colors.white,
      hover: colors.white,
      disabled: colors.white,
    },
  },
  white: {
    fontColor: {
      default: indigoPalette.primary,
      hover: indigoPalette.primary,
      disabled: indigoPalette.lighter,
    },
    shadow: {
      hover: DEFAULT_SHADOW,
    },
    outline: {
      bg: {
        default: indigoPalette.primary,
        hover: indigoPalette.primary,
        disabled: indigoPalette.primary,
      },
      border: {
        disabled: indigoPalette.light,
      },
      fontColor: {
        default: colors.white,
        hover: colors.white,
        disabled: indigoPalette.lighter,
      },
      shadow: {
        hover: DEFAULT_SHADOW,
      },
    },
  },
  secondary: {
    bg: {
      hover: hackBluePalette.darker,
      disabled: impactGreyPalette.light,
    },
    fontColor: {
      disabled: impactGreyPalette.midtone,
    },
    outline: {
      bg: {
        disabled: colors.white,
      },
      border: {
        hover: hackBluePalette.darker,
        disabled: impactGreyPalette.light,
      },
      fontColor: {
        disabled: indigoPalette.lighter,
      },
    },
  },
};


const link = {
  default: (colorPalette) => ({
    color: {
      default: colorPalette.primary,
      hover: colorPalette.dark,
    },
  }),
  secondary: {
    default: hackBluePalette.primary,
    hover: hackBluePalette.darker,
  },
};


const theme = {
  // General
  colors,
  fonts,
  fontWeights,
  fontSizes,
  button,
  link,
  DEFAULT_SHADOW,
};


export default theme;
export { colors };
