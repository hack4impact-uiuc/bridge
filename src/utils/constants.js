import * as styledSystem from 'styled-system';
import systemPropTypes from '@styled-system/prop-types';
import PropTypes from 'prop-types';


const { compose, system } = styledSystem;

// Text Decoration Prop
export const textDecoration = system({
  prop: 'textDecoration',
  cssProperty: 'text-decoration',
});
textDecoration.propTypes = PropTypes.string;


// common style props to be passed into components
export const COMMON = compose(styledSystem.space, styledSystem.layout);
COMMON.propTypes = {
  ...systemPropTypes.space,
  ...systemPropTypes.layout,
};

// for easier access
export const TYPOGRAPHY = styledSystem.typography;
export const LAYOUT = styledSystem.layout;
export const POSITION = styledSystem.position;
export const FLEX = styledSystem.flexbox;
export const GRID = styledSystem.grid;
export const COLOR = styledSystem.color;
export const BORDER = styledSystem.border;

TYPOGRAPHY.propTypes = systemPropTypes.typography;
LAYOUT.propTypes = systemPropTypes.layout;
POSITION.propTypes = systemPropTypes.position;
FLEX.propTypes = systemPropTypes.flexbox;
GRID.propTypes = systemPropTypes.grid;
COLOR.propTypes = systemPropTypes.color;
BORDER.propTypes = systemPropTypes.border;
