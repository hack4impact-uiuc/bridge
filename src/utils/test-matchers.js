/* eslint-disable react/forbid-foreign-prop-types */
// taken from https://github.com/primer/components/blob/master/src/utils/test-matchers.js
import 'jest-styled-components';

const stringify = (d) => JSON.stringify(d, null, '  ');
/**
 * These are props that styled-system aliases for backwards compatibility.
 * For some reason, they don't show up in our toImplementSystemProps() matcher,
 * so we skip over them.
 */
const ALIAS_PROP_TYPES = ['w', 'align', 'justify', 'wrap'];

// eslint-disable-next-line no-undef
expect.extend({
  toImplementSystemProps(Component, propNames) {
    const propKeys = new Set(Object.keys(Component.propTypes));
    const expectedPropKeys = Object.keys(propNames.propTypes);
    const missing = expectedPropKeys.filter((key) => !propKeys.has(key))
      .filter((key) => !ALIAS_PROP_TYPES.includes(key));
    return {
      pass: missing.length === 0,
      message: () => `Missing prop${missing.length === 1 ? '' : 's'}: ${stringify(missing)}`,
    };
  },
});
