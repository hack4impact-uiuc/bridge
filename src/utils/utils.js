import themeGet from '@styled-system/theme-get';
import * as styledSystem from 'styled-system';
import theme from '../theme';

const { get: getKey } = styledSystem;

export const get = (key) => themeGet(key, getKey(theme, key));

export const getOrDefault = (val, def) => (val !== undefined && val !== null ? val : def);

export const lodashGet = (object, path, defaultVal) => {
  const PATH = Array.isArray(path)
    ? path
    : path.split('.').filter((i) => i.length);
  if (!PATH.length) {
    return object === undefined ? defaultVal : object;
  }
  if (object === null || object === undefined || typeof (object[PATH[0]]) === 'undefined') {
    return defaultVal;
  }
  return lodashGet(object[PATH.shift()], PATH, defaultVal);
};
