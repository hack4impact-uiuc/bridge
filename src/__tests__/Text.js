/* eslint-disable jsx-a11y/anchor-is-valid */
import React from 'react';
import { render, cleanup } from '@testing-library/react';
import 'jest-styled-components';
import '@testing-library/jest-dom';
import { axe, toHaveNoViolations } from 'jest-axe';

import { Text } from '..';
import theme from '../theme';
import { COMMON, TYPOGRAPHY } from '../utils/constants';
import { renderJSON } from '../utils/testing';
import 'babel-polyfill'; // axe violations required babel-polyfill

expect.extend(toHaveNoViolations);

describe('Text', () => {
  it('should have no axe violations', async () => {
    const { container } = render(<Text as="p">APPLY</Text>);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
    cleanup();
  });

  it('implements system props', () => {
    expect(Text).toImplementSystemProps(TYPOGRAPHY);
    expect(Text).toImplementSystemProps(COMMON);
  });

  it('renders a <span> by default', () => {
    expect(renderJSON(<Text />).type).toEqual('span');
  });

  it('respects fontWeight', () => {
    const { container } = render(<Text fontWeight="bold" />);
    expect(container.firstChild).toHaveStyle(`font-weight: ${theme.fontWeights.bold}px`);

    const { container: c } = render(<Text fontWeight="regular" />);
    expect(c.firstChild).toHaveStyle(`font-weight: ${theme.fontWeights.regular}px`);
  });

  it('respects the "fontStyle" prop', () => {
    const { container } = render(<Text fontStyle="italic" />);
    expect(container.firstChild).toHaveStyle('font-style: italic');

    const { container: c } = render(<Text as="i" fontStyle="normal" />);
    expect(c.firstChild).toHaveStyle('font-style: normal');
  });

  it('respects the "as" prop', () => {
    expect(renderJSON(<Text as="p" />).type).toEqual('p');
  });

  it('respects the "as" prop with non-string component types', () => {
    // eslint-disable-next-line react/prop-types
    function TestP({ children }) {
      return <p className="test-link">{children}</p>;
    }

    expect(renderJSON(<Text as={TestP}>Testing...</Text>)).toMatchSnapshot();
  });

  it('respects lineHeight', () => {
    const { container } = render(<Text lineHeight="80px" />);
    expect(container.firstChild).toHaveStyle('line-height: 80px}');
  });

  it('respects fontFamily="secondary"', () => {
    // styled-components removes the whitespace between font-family values
    const sec = theme.fonts.secondary.replace(/, /g, ',');
    const { container } = render(<Text fontFamily="secondary" />);
    expect(container.firstChild).toHaveStyle(`font-family: ${sec}`);
  });

  it('respects other values for fontSize', () => {
    const { container } = render(<Text fontSize="2em" />);
    expect(container.firstChild).toHaveStyle('font-size: 2em');

    const { container: c } = render(<Text fontSize="100px" />);
    expect(c.firstChild).toHaveStyle('font-size: 100px');
  });
});

describe('Text types', () => {
  it('defaults to type="body"', () => {
    expect(renderJSON(<Text>Hello</Text>)).toMatchSnapshot();
    const { container } = render(<Text>Hello</Text>);
    expect(container.firstChild).toHaveStyle(`font-family: ${theme.typography.body.fontFamily}`);
    expect(container.firstChild).toHaveStyle(`font-weight: ${theme.typography.body.fontWeight}px`);
    expect(container.firstChild).toHaveStyle(`font-size: ${theme.typography.body.fontSize}`);
    expect(container.firstChild).toHaveStyle(`letter-spacing: ${theme.typography.body.letterSpacing}`);
    expect(container.firstChild).toHaveStyle(`line-height: ${theme.typography.body.lineHeight}`);
  });

  it('renders "hero" properly', () => {
    expect(renderJSON(<Text type="hero">Hello</Text>)).toMatchSnapshot();
  });

  it('renders "overline" properly', () => {
    expect(renderJSON(<Text type="overline">Hello</Text>)).toMatchSnapshot();
  });

  it('renders "button" properly', () => {
    expect(renderJSON(<Text type="button">Hello</Text>)).toMatchSnapshot();
  });

  it('renders "buttonSmall" properly', () => {
    expect(renderJSON(<Text type="buttonSmall">Hello</Text>)).toMatchSnapshot();
  });

  it('renders "quote" properly', () => {
    expect(renderJSON(<Text type="quote">Hello</Text>)).toMatchSnapshot();
  });

  it('renders "caption" properly', () => {
    expect(renderJSON(<Text type="caption">Hello</Text>)).toMatchSnapshot();
  });

  it('renders "table" properly', () => {
    expect(renderJSON(<Text type="table">Hello</Text>)).toMatchSnapshot();
  });

  it('renders "alert" properly', () => {
    expect(renderJSON(<Text type="alert">Hello</Text>)).toMatchSnapshot();
  });

  it('renders "form" properly', () => {
    expect(renderJSON(<Text type="form">Hello</Text>)).toMatchSnapshot();
  });

  it('renders "subtitle" properly', () => {
    expect(renderJSON(<Text type="subtitle">Hello</Text>)).toMatchSnapshot();
  });
});
