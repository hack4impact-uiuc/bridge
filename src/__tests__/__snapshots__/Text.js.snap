// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text respects the "as" prop with non-string component types 1`] = `
<p
  className="test-link"
>
  Testing...
</p>
`;

exports[`Text types defaults to type="body" 1`] = `
.c0 {
  color: #0A162A;
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 16px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 24px;
}

<span
  className="c0"
  type="body"
>
  Hello
</span>
`;

exports[`Text types renders "alert" properly 1`] = `
.c0 {
  color: #0A162A;
}

<span
  className="c0"
  type="alert"
>
  Hello
</span>
`;

exports[`Text types renders "button" properly 1`] = `
.c0 {
  color: #0A162A;
  font-family: Chivo;
  font-weight: 700px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  font-size: 14px;
  line-height: 20px;
}

<span
  className="c0"
  type="button"
>
  Hello
</span>
`;

exports[`Text types renders "buttonSmall" properly 1`] = `
.c0 {
  color: #0A162A;
  font-family: Chivo;
  font-weight: 700px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  font-size: 12px;
  line-height: 16px;
}

<span
  className="c0"
  type="buttonSmall"
>
  Hello
</span>
`;

exports[`Text types renders "caption" properly 1`] = `
.c0 {
  color: #0A162A;
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 14px;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  line-height: 20px;
  font-style: italic;
}

<span
  className="c0"
  type="caption"
>
  Hello
</span>
`;

exports[`Text types renders "form" properly 1`] = `
.c0 {
  color: #0A162A;
}

<span
  className="c0"
  type="form"
>
  Hello
</span>
`;

exports[`Text types renders "hero" properly 1`] = `
.c0 {
  color: #0A162A;
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 22px;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  line-height: 32px;
}

<span
  className="c0"
  type="hero"
>
  Hello
</span>
`;

exports[`Text types renders "overline" properly 1`] = `
.c0 {
  color: #0A162A;
  font-family: HKGrotesk;
  font-weight: 700px;
  font-size: 12px;
  -webkit-letter-spacing: 2px;
  -moz-letter-spacing: 2px;
  -ms-letter-spacing: 2px;
  letter-spacing: 2px;
  line-height: 16px;
  text-transform: uppercase;
}

<span
  className="c0"
  type="overline"
>
  Hello
</span>
`;

exports[`Text types renders "quote" properly 1`] = `
.c0 {
  color: #0A162A;
  font-family: HKGrotesk;
  font-weight: 300px;
  font-size: 28px;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  line-height: 40px;
}

<span
  className="c0"
  type="quote"
>
  Hello
</span>
`;

exports[`Text types renders "subtitle" properly 1`] = `
.c0 {
  color: #0A162A;
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 20px;
  -webkit-letter-spacing: 0.5px;
  -moz-letter-spacing: 0.5px;
  -ms-letter-spacing: 0.5px;
  letter-spacing: 0.5px;
  line-height: 28px;
}

<span
  className="c0"
  type="subtitle"
>
  Hello
</span>
`;

exports[`Text types renders "table" properly 1`] = `
.c0 {
  color: #0A162A;
  font-family: HKGrotesk;
  font-weight: 400px;
  font-size: 18px;
  -webkit-letter-spacing: 0.3px;
  -moz-letter-spacing: 0.3px;
  -ms-letter-spacing: 0.3px;
  letter-spacing: 0.3px;
  line-height: 24px;
}

<span
  className="c0"
  type="table"
>
  Hello
</span>
`;
